toolchain/check/import_ref.cpp

```diff
@@ -258,13 +284,47 @@ class ImportRefResolver {
     return import_name_id;
   }
 
+  // Translates a NameScopeId from the import IR to a local NameScopeId. Adds
+  // unresolved constants to the work stack.
+  auto GetLocalNameScopeId(SemIR::NameScopeId name_scope_id)
+      -> SemIR::NameScopeId {
+    if (!name_scope_id.is_valid()) {
+      return SemIR::NameScopeId::Invalid;
+    }
+    auto inst_id = import_ir_.name_scopes().Get(name_scope_id).inst_id;
+    if (!inst_id.is_valid()) {
+      // Map scopes that aren't associated with an instruction to invalid
+      // scopes. For now, such scopes aren't used, and we don't have a good way
+      // to rmmap them.
+      return SemIR::NameScopeId::Invalid;
+    }
+    auto const_id = GetLocalConstantId(inst_id);
+    if (!const_id.is_valid()) {
+      return SemIR::NameScopeId::Invalid;
+    }
+    auto name_scope_inst = context_.insts().Get(const_id.inst_id());
+    if (auto namespace_inst = name_scope_inst.TryAs<SemIR::Namespace>()) {
+      return namespace_inst->name_scope_id;
+    }
+    if (auto class_inst = name_scope_inst.TryAs<SemIR::ClassType>()) {
+      return context_.classes().Get(class_inst->class_id).scope_id;
+    }
+    if (auto interface_inst = name_scope_inst.TryAs<SemIR::InterfaceType>()) {
+      return context_.interfaces().Get(interface_inst->interface_id).scope_id;
+    }
```

Comment:
Should this be a switch on the kind rather than repeated cast attempts? I think it might be easier to read, but couldn't the repeated TryAs also be a performance issue, particularly if we add more scoped entities?

Reply:
Done.

While I actually find the result a little harder to read and more error-prone due to writing each type twice rather than once (so I marginally prefer the old form in isolation), I think I'd prefer a `switch` once we add a fourth case here, which we probably will :)

FWIW, I don't think performance is a factor here. After inlining, this'll just be repeated comparisons against the kind, which the optimizer should canonicalize into the same form as the `switch` anyway. (And at `-O0` a `switch` over a few cases gets lowered into a series of `if`s. Though we would avoid building the `optional`s at `-O0`.)

Reply:
FWIW, I kind of agree on the redundancy, but I find the switch a lot easier to see consistencies in -- the `if (...)` block I spend time trying to figure out the behavior. For example versus the `TryAs` code, I can more trivially see the consistent `return` behavior, and the cross-case flow is enforced by `switch` (and we do have things like the `Param` `TryAs` uses which do more of a fallthrough behavior, emphasizing the need to check). The `default` handling is similarly clearer for me.

If we did more macros, it'd kind of be tempting to write something that encapsulated:

```
#define CARBON_INST_CASE(InstName) \
case InstName::Kind: \
   auto typed_inst = inst.As<InstName>();
```

Reply:
Hmmm, I think we could probably get:

```
TypedInstSwitch(context_.insts().Get(const_id.inst_id()))
    .Case<SemIR::Namespace>([](auto inst) { return inst.name_scope_id; })
    .Case<SemIR::ClassType>([](auto inst) { return context_.classes().Get(inst.class_id).scope_id; })
etc
```

Is this worthwhile, do you think?

Reply:
Per discussion, I'm going to try this out.